test/common/http/codec_impl_fuzz_test.cc

```diff
@@ -671,6 +675,9 @@ void codecFuzz(const test::common::http::CodecImplFuzzTestCase& input, HttpVersi
           continue;
         }
       }
+
+      const auto allow_h1_trailers = input.h1_settings().server().enable_trailers() &&
```

Comment:
I wonder whether this should be:
`const auto h1_send_trailers = input.h1_settings().server().enable_trailers() == input.h1_settings().client().enable_trailers();`, i.e., sending trailers either when both support it or both don't support it. WDYT?

Reply:
Unfortunately does `const auto h1_allow_trailers = input.h1_settings().server().enable_trailers() == input.h1_settings().client().enable_trailers();` still trigger the assert, when both are false. I am not sure whether I found an actual bug in envoy here, where the support for h1 trailers is buggy. What do you think?

Please note that I did intentionally name the variable `h1_allow_trailers`. This is to emphasize that sending trailers make sense only, when both server and client agree on this.

Reply:
I don't think this is an envoy bug. The original issue is we were encoding trailers when telling the codec we weren't supporting trailers. I'm fine with these being coupled right now (e.g. trailers either on or off).

I think to add support for them to occur independently we'd want to push down the logic here: 

https://github.com/envoyproxy/envoy/pull/21124/files/cac67cb8ef41bd055ff12af0cb2970e96f57f6d4#diff-d376dede7065839b9909c5d38020a189e9024bdfac250cbbacf0580fb654825aR299-R305

To only encodeTrailers for request if the client support trailers and only encodeTrailer for response if server support trailers, though it might require more tuning.

Reply:
Analyzing this further I more and more believe this is a severe bug. When on an H1 stream trailers are send without it being configured to enable trailers, then `StreamEncoderImpl::endEncode()` is called by `StreamEncoderImpl::encodeTrailersBase()` which does not respect any `ClientConnectionImpl::pending_response_`'s set. Deactivating the `h1_allow_trailers` flag again and patching envoy as in https://github.com/envoyproxy/envoy/pull/21124/commits/782f3a665fe52b19a427e939168e531d573e6445 makes it accept trailers and behave like I assume it should, but with this change it crashes so many other tests, that I had to revert it immediately.

I am now totally frustrated giving up on this issue. I assume, that I know what should be done, but patching envoy to get it to handle trailers correctly is beyond me.

Reply:
/ping
How do we proceed with this issue?

Reply:
I've discussed this offline with @KBaichoo. [Here](https://docs.google.com/document/d/1WzXNZA9diZCasLcqaaiCtZErgh3-IBRqP8CDkqCk2cI/edit#) is a short explanation (with a diagram) that I think will make things a bit clearer to discuss.

Reply:
Well, when I understand you correctly in that doc, then `request_.request_decoder_.decodeTrailers_(_)` being a mock shall be filtering any trailers for h1 where no trailers are enabled, right? I.e. in the mock [`ON_CALL( of the above )`](https://github.com/envoyproxy/envoy/blob/b939aa2fc166f87b65a2aa439d0445f3f86f18fb/test/common/http/codec_impl_fuzz_test.cc#L198) some magic has to be wrought, that ... well what? Terminates processing? Filters trailers? Or what?

Whatever there should be done, has the problem, that the mock-function does not get called (note I have commented in line [297](https://github.com/envoyproxy/envoy/blob/b939aa2fc166f87b65a2aa439d0445f3f86f18fb/test/common/http/codec_impl_fuzz_test.cc#L297) the h1-trailer-checks to allow for calling this routine, but the debugger will **not** break there). So either there is an order issue (where the dispatching of the old stream and creation of the new on must not overlap) or the mock-functions are not called as expected.

Reply:
/ping  @adisuissa @KBaichoo 
How do we proceed?

Reply:
Hey @vehre-x41 , sorry for the delay. Here's the PR for this that runs both test cases and independently supports trailers. A little rough around the edges (will clean up): https://github.com/envoyproxy/envoy/pull/21730